### ISSUE TITLE
Optimize "DeleteGuests" stored procedure

### ISSUE THREAD

**Issue by AndreiMaz** [nopSolutions/nopCommerce#4087]
Source: https://www.nopcommerce.com/boards/t/72025/stored-procedure-deleteguests-cross-join-.aspx

**Comment by Leftyx** [comment]
I noticed this today cause the stored procedure `[DeleteGuests] `was crashing with timeout.
A good T-SQL course for these guys would help a lot.

**Comment by Leftyx** [comment]
There you go
```

CREATE PROCEDURE [dbo].[DeleteGuests]
(
	@OnlyWithoutShoppingCart bit = 1,
	@CreatedFromUtc datetime,
	@CreatedToUtc datetime,
	@TotalRecordsDeleted int = null OUTPUT
)
AS
BEGIN
	CREATE TABLE #tmp_guests (CustomerId int)
		
	INSERT #tmp_guests (CustomerId)
	SELECT [Id] FROM [Customer] c with (NOLOCK)
	WHERE
	--created from
	((@CreatedFromUtc is null) OR (c.[CreatedOnUtc] > @CreatedFromUtc))
	AND
	--created to
	((@CreatedToUtc is null) OR (c.[CreatedOnUtc] < @CreatedToUtc))
	AND
	--shopping cart items
	((@OnlyWithoutShoppingCart=0) OR (NOT EXISTS(SELECT 1 FROM [ShoppingCartItem] sci with (NOLOCK) where sci.[CustomerId]=c.[Id])))
	AND
	--guests only
	(EXISTS(SELECT 1 FROM [Customer_CustomerRole_Mapping] ccrm with (NOLOCK) inner join [CustomerRole] cr with (NOLOCK) on cr.[Id]=ccrm.[CustomerRole_Id] WHERE ccrm.[Customer_Id]=c.[Id] AND cr.[SystemName] = N'Guests'))
	AND
	--no orders
	(NOT EXISTS(SELECT 1 FROM [Order] o with (NOLOCK) WHERE o.[CustomerId]=c.[Id]))
	AND
	--no blog comments
	(NOT EXISTS(SELECT 1 FROM [BlogComment] bc with (NOLOCK) WHERE bc.[CustomerId]=c.[Id]))
	AND
	--no news comments
	(NOT EXISTS(SELECT 1 FROM [NewsComment] nc with (NOLOCK) WHERE nc.[CustomerId]=c.[Id]))
	AND
	--no product reviews
	(NOT EXISTS(SELECT 1 FROM [ProductReview] pr with (NOLOCK) WHERE pr.[CustomerId]=c.[Id]))
	AND
	--no product reviews helpfulness
	(NOT EXISTS(SELECT 1 FROM [ProductReviewHelpfulness] prh with (NOLOCK) WHERE prh.[CustomerId]=c.[Id]))
	AND
	--no poll voting
	(NOT EXISTS(SELECT 1 FROM [PollVotingRecord] pvr with (NOLOCK) WHERE pvr.[CustomerId]=c.[Id]))
	AND
	--no forum topics 
	(NOT EXISTS(SELECT 1 FROM [Forums_Topic] ft with (NOLOCK) WHERE ft.[CustomerId]=c.[Id]))
	AND
	--no forum posts 
	(NOT EXISTS(SELECT 1 FROM [Forums_Post] fp with (NOLOCK) WHERE fp.[CustomerId]=c.[Id]))
	AND
	--no system accounts
	(c.IsSystemAccount = 0)
	
	--delete guests
	DELETE [Customer]
	WHERE [Id] IN (SELECT [CustomerId] FROM #tmp_guests)
	
	--delete attributes
	DELETE [GenericAttribute]
	WHERE ([EntityId] IN (SELECT [CustomerId] FROM #tmp_guests))
	AND
	([KeyGroup] = N'Customer')
	
	--total records
	SELECT @TotalRecordsDeleted = COUNT(1) FROM #tmp_guests
	
	DROP TABLE #tmp_guests
END

GO
```

**Comment by AndreiMaz** [comment]
After testing on our official website I can confirm that the above stored procedure works much faster

@Leftyx Thanks a lot

**Comment by Leftyx** [comment]
@AndreiMaz : I am sorry if I have been a little mean in my comments and in the past but I think you guys need to work on performances quite a lot ... and spend some time on it.

NopCommerce works really well with a bunch of products and a few categories with no localization but if you start pushing in data it starts to struggle.  
Of course there's always the option to add more CPUs and memory but I guess with .NET core things should scale quite well.  
I Have been working on NopCommerce for quite a while and have tried to fix some of the issues. I'd like to contribute but some of the changes would  be rejected, surely, cause they are a complete overhaul of the product itself.  

I'll do my best to contribute to the project.

**Comment by holydk** [comment]
Closed #4087
